src/sm.c

```diff
@@ -0,0 +1,331 @@
+{
+    SM_RESULT result;
+    /*Codes_SRS_SM_02_021: [ If sm is NULL then sm_begin shall fail and return SM_ERROR. ]*/
+    if (sm == NULL)
+    {
+        LogError("invalid argument SM_HANDLE sm=%p", sm);
+        result = SM_ERROR;
+    }
+    else
+    {
+        LONG64 b_now_1 = InterlockedAdd64(&sm->b_now, 0);
+
+        /*Codes_SRS_SM_02_022: [ If there's a barrier set then sm_begin shall return SM_EXEC_REFUSED. ]*/
+        if ((b_now_1 & 1)==1)
+        {
+            result = SM_EXEC_REFUSED;
+        }
+        else
+        {
+            /*Codes_SRS_SM_02_035: [ sm_begin shall increment n. ]*/
+            (void)InterlockedIncrement64(&sm->n);
+
+            LONG64 b_now_2 = InterlockedAdd64(&sm->b_now, 0);
+
+            /*Codes_SRS_SM_02_036: [ If the barrier changed after incrementing n then sm_begin shall increment e, signal a potential drain, and return SM_EXEC_REFUSED. ]*/
+            if (b_now_2 != b_now_1)
+            {
+                LONG64 e = InterlockedIncrement64(&sm->e);
+                LONG64 n = InterlockedAdd64(&sm->n, 0);
+                if (n - e == 1)
+                {
+                    InterlockedHL_SetAndWake((void*)&sm->e_done, 1);
+                }
+                result = SM_EXEC_REFUSED;;
+            }
+            else
+            {
+                /*Codes_SRS_SM_02_023: [ sm_begin shall succeed and return SM_EXEC_GRANTED. ]*/
+                result = SM_EXEC_GRANTED;
+            }
+        }
+    }
+    
+    return result;
+}
+
+void sm_end(SM_HANDLE sm)
+{
+    SM_RESULT result;
+    /*Codes_SRS_SM_02_024: [ If sm is NULL then sm_end shall return. ]*/
+    if (sm == NULL)
+    {
+        LogError("invalid argument SM_HANDLE sm=%p", sm);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        /*Codes_SRS_SM_02_025: [ sm_end shall increment the number of executed APIs (e). ]*/
+        LONG64 e = InterlockedIncrement64(&sm->e);
+
+        /*Codes_SRS_SM_02_026: [ If n-e is 1 then sm_end shall wake up the waiting barrier. ]*/
+        LONG64 n = InterlockedAdd64(&sm->n, 0);
+        if (n - e == 1)
+        {
+            InterlockedHL_SetAndWake((void*)&sm->e_done, 1);
+        }
+    }
+}
+
+SM_RESULT sm_barrier_begin(SM_HANDLE sm)
+{
+    SM_RESULT result;
+    /*Codes_SRS_SM_02_027: [ If sm is NULL then sm_barrier_begin shall fail and return SM_ERROR. ]*/
+    if (sm == NULL)
+    {
+        LogError("invalid argument SM_HANDLE sm=%p", sm);
+        result = SM_ERROR;
+    }
+    else
+    {
+        result=sm_barrier_begin_internal(sm);
+    }
+
+    return result;
+}
+
+void sm_barrier_end(SM_HANDLE sm)
+{
+    /*Codes_SRS_SM_02_032: [ If sm is NULL then sm_barrier_end shall return. ]*/
+    if (sm == NULL)
+    {
+        LogError("invalid argument SM_HANDLE sm=%p", sm);
+    }
+    else
+    {
+        /*Codes_SRS_SM_02_033: [ sm_barrier_end shall increment the number of executed operations (e), increment b_now and return. ]*/
+        InterlockedIncrement64(&sm->e);
```

Comment:
>InterlockedIncrement64 [](start = 8, length = 22)

(void)

Reply:
Here and below.

---
In reply to: [421818062](https://github.com/Azure/azure-c-util/pull/35#discussion_r421818062) [](ancestors = 421818062)